### PR TITLE
convlstm post (now ready for review)

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+
+Now, about that public vs. private distinction. In both frameworks, we can obtain outputs (hidden states) for every time step. The cell state, however, we can access only for the very last time step. This, it need be said, is purely an implementation decision. As we'll see when building our own recurrent module, there are no obstacles inherent in keeping track of cell states and passing them back to the user.
+
+If you dislike the pragmatism of this distinction, you can always go with the math. When a new cell state has been computed (based on prior cell state, input, forget, and cell gates -- the specifics of which we are not going to get into here), it is transformed to the hidden (a.k.a. output) state making use of yet another, the output gate:
+
+$$
+h_t = o_t \odot \tanh(c_t)
+$$
+
+Definitely, then, hidden state (output, resp.) builds on cell state, adding additional modeling power. Whether one would interpret the workings of the output gate in an anthropomorphizing way is left to personal choice.
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:161]: This is probably because I don't understand the technical aspects, but I'm not even sure how one would go about interpreting the output gate as though it were human. If you think this last sentence adds something, then you can leave it. But, given the amount of prose in this piece (all prior to actually building the convLSTM), I'd consider cutting it.

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+: Table 2: Contrasting ways of obtaining various return values in `torch` vs. Keras. Cf. the appendix for complete examples.
+:::
+
+<br />
+
+### `convLSTM`, the plan
+
+In both `torch` and Keras RNN architectures, single time steps are processed by corresponding `Cell` classes: There is an LSTM Cell matching the LSTM, a GRU Cell matching the GRU, and so on. We do the same for ConvLSTM. In `convlstm_cell()`, we first define what should happen to a single observation; then in `convlstm()`, we build up the recurrence logic.
+
+Once we're done, we create a dummy dataset, as reduced-to-the-essential as can be. With more complex datasets, even artificial ones, chances are that if we don't see any training progress, there are hundreds of possible explanations. We want a sanity check that, if we fail it, leaves no excuses. Realistic applications are left to future posts.
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:189]:
```suggestion
Once we're done, we create a dummy dataset, as reduced-to-the-essentials as can be. With more complex datasets, even artificial ones, chances are that if we don't see any training progress, there are hundreds of possible explanations. We want a sanity check that, if failed, leaves no excuses. Realistic applications are left to future posts.
```

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+
+### A single step: `convlstm_cell`
+
+Our `convlstm_cell`'s constructor takes arguments `input_dim` , `hidden_dim`, and `bias`, just like a `torch` LSTM Cell.
+
+But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution, of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:195]:
```suggestion
But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.
```

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 4]
```diff
@@ -0,0 +1,753 @@
...
+
+Note how the `channels` dimension, which in the original input data would correspond to different variables, is creatively used to consolidate four convolutions into one: Each channel output will be passed to just one of the four cell gates. Once in possession of the convolution output, `forward()` applies the gate logic, resulting in the two types of states it needs to send back to the caller.
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:197]: This paragraph is ⭐️!

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+# 2  1 16 16
+```
+
+Now we want to sanity-check this module with the simplest-possible dummy data.
+
+### Sanity-checking the `convlstm`
+
+We generate black-and-white "movies" of diagonal beams successively translated in space.
+
+Each sequence consists of six time steps, and each beam, of six pixels. Just a single sequence is created manually. To create that one sequence, we start from a single beam:
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:406]:
```suggestion
Each sequence consists of six time steps, and each beam of six pixels. Just a single sequence is created manually. To create that one sequence, we start from a single beam:
```

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 6]
```diff
@@ -0,0 +1,753 @@
...
+     [5,]  0.00 -0.01 -0.01 -0.01  0.75  0.01  0.06  0.00  0.00     0
+     [6,]  0.00  0.01  0.00 -0.07 -0.01  0.75  0.01  0.06  0.00     0
+     [7,]  0.00  0.01 -0.01 -0.01 -0.07 -0.01  0.75  0.01  0.06     0
+     [8,]  0.00  0.00  0.01  0.00  0.00 -0.01  0.00  0.71  0.00     0
+     [9,]  0.00  0.00  0.00  0.01  0.01  0.00  0.03 -0.01  0.37     0
+    [10,]  0.00  0.00  0.00  0.00  0.00  0.00 -0.01 -0.01 -0.01     0
+
+This should suffice for a sanity check. If you made it till the end, thanks for your patience! In the best case, you'll be able to apply this architecture (or a similar one) to your own data -- but even if not, I hope you've enjoyed learning about `torch` model coding and/or RNN weirdness ;-)
+
+Myself, I am certainly looking forward to exploring convLSTMs on real-world problems in the near future. Thanks for reading!
```
Comment (batpigandme) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:558]: Myself is a reflexive pronoun, which you only use when it's the object of the sentence. The best way I can think of it is with french verbs, _se laver_ to wash (you wash yourself). You can put yourself to bed, bathe yourself, kill yourself…there aren't a lot of reflexive verbs in English (or at least I can't think of them right now). Basically, people use myself incorrectly all the time (including above). Myself can also be used as an intensive pronoun, but, in that case, it should go after the "I"— _I myself am certainly looking forward to…_

**Review by batpigandme — APPROVED** [review]
This is a long one, and I've left it deliberately chatty (as I'm pretty sure that's what's intended). Most edits are minor—just make sure that the reader doesn't get so lost in the asides that they can't resolve the sentence.

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 7]
```diff
@@ -0,0 +1,753 @@
...
+
+What do I expect? Let's approach this task in a top-down way.
+
+### Input and output
+
+The convLSTM's input will be a time series of spatial data, each observation being of size `(time steps, channels, height, width)`.
+
+Compare this with the usual RNN input format, be it in `torch` or Keras. In both frameworks, RNNs expect tensors of size `(timesteps, input_dim)`[^1]. `input_dim` is $1$ for univariate time series and greater than $1$ for multivariate ones. Conceptually, we may match this to convLSTM's `channels` dimension: There could be a single channel, for temperature, say -- or there could be several, such as for pressure, temperature, and humidity. The two additional dimensions found in convLSTM, `height` and `width`, are spatial indexes into the data.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:54]: lol :-) I tried again ;-) ... I really prefer `indexes` (but that may be because I had several years (((working as a DBA))) to digest it ;-)))

Re technical language

> Meanwhile, indices is generally preferred in mathematical, financial, and technical contexts, while indexes is relatively common in general usage.

I'm not sure how they mean it, but personally I'd only ever use indices when referring to i, j, k or the like, but otherwise ... like in the above sentence ... I think indexes sounds more natural ...

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+
+We want our convLSTM to be usable in different contexts and applications. Various architectures exist that make use of hidden states, most prominently perhaps, encoder-decoder architectures. Thus, we want our convLSTM to return those as well. Again, this is something a `torch` LSTM does by default, while in Keras it is achieved using `return_state = TRUE`.
+
+Now though, it really is time for that interlude. We'll sort out the ways things are called by both `torch` and Keras, and inspect what you get back from their respective GRUs and LSTMs.
+
+### Interlude: Outputs, states, hidden values ... what's what?
+
+For this to remain an interlude, I intend to summarize findings on a high level. The code snippets in the appendix show how to arrive at these results. Heavily commented, they probe return values from both Keras and `torch` GRUs and LSTMs. Running these will make the upcoming summaries seem a lot less abstract.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:78]: lol :-) don't force me to say ;-)

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+
+Now, *about those hidden states*.
+
+When a Keras LSTM is defined with `return_state = TRUE`, its return value is a structure of three entities called output, memory state, and carry state. In `torch`, the same entities are referred to as output, hidden state, and cell state. (In `torch`, we always get all of them.)
+
+So are we dealing with three different types of entities? We are not.
+
+The cell, or carry, state is that special thing setting apart LSTMs from GRUs deemed responsible for the "long" in "long short-term memory". Technically, it could be reported to the user at all points in time; as we'll see shortly though, it is not.
+
+What about outputs and hidden, or memory, states? Confusingly, these really are the same thing. Recall that for each input *item* in the input *sequence*, we're combining it with the previous state, resulting in a new state, to be made used of in the next step[^3]:
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:113]: interesting - I didn't know!

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:197]: hehe thanks :-))

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 10]
```diff
@@ -0,0 +1,753 @@
+---
+title: "Convolutional LSTM for spatial forecasting"
+description: > 
+  In forecasting spatially-determined phenomena (the weather, say, or the next frame in a movie), we want to model temporal evolution, ideally using recurrence relations, while efficiently extracting spatial features, as usually done using convolutions. Ideally, we'd have at our disposal an architecture that is both recurrent and convolutional. In this post, we build a convolutional LSTM with torch.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:4]: backticks don't work in the abstract ... but otherwise ... yeah ... unfortunately all that is correct :-(
will see what I can do

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+preview: images/preview.jpeg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+This post is the first in a loose series aimed at exploring forecasting of spatially-determined data over time. By spatially-determined I mean that whatever the quantities we're trying to predict -- be they univariate or multivariate time series, of spatial dimensionality or not -- the *input* data are given on a spatial grid.
+
+For example, the input could be atmospheric measurements, such as sea surface temperature or pressure, given at some set of latitudes and longitudes. The target to be predicted could then span that same (or another) grid. Alternatively, it could be a univariate time series, like a meteorological index.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:30]: yes :-)

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+
+But wait a second, you may be thinking. For time series prediction, we have that time-honored set of recurrent architectures, right -- LSTM, GRU ...? Right. We do; but once we feed spatial data to an RNN, treating different locations as different input features, we lose an essential structural relationship. Importantly, we need to operate in both space and time. We want both: recurrence relations and convolutional filters. Enter *convolutional RNNs*.
+
+## What to expect from this post
+
+Today, we won't jump into real-world applications just yet. Instead, we'll take our time to build a convolutional LSTM (henceforth: convLSTM) in `torch`. For one, we have to -- there is no official PyTorch implementation. (As opposed to Keras, by the way. If you're interested in quickly playing around with a Keras convLSTM, check out [this nice example](https://tensorflow.rstudio.com/guide/keras/examples/conv_lstm/).)
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:36]: great idea!

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:161]: you're right as always - this is a leftover hinting at an -  even longer - paragraph that got completely removed (because verbosity ... yeah ....)

**Comment by skeydan** [comment]
Thanks a 1000 times @batpigandme for the awesome review!!

Also, am feeling flattered coz you call me chatty, which probably does not happen too often ;-)))

(But yeah, I see what you mean ... but splitting this up did not seem like a good idea)

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 13]
```diff
@@ -0,0 +1,757 @@
...
+```
+
+That's it for the raw data. Now we still need a `dataset` and a `dataloader`. Of the six time steps, we use the first five as input and try to predict the last one.
+
+```{r}
+dummy_ds <- dataset(
+  
+  initialize = function(data) {
+    # add channels dimension
+    self$data <- data$unsqueeze(3)
```
Comment (dfalbel) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:452]: i found it confusing for this to be there. I would make the `input` sequence with the correct size isntead.

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 14]
```diff
@@ -0,0 +1,757 @@
...
+
+# for each layer, tensor of size (batch_size, seq_len, hidden_size, height, width)
+dim(layer_outputs[[1]])
+# 2  4  5 16 16
+dim(layer_outputs[[3]])
+# 2  4  1 16 16
+
+# list of 2 tensors for each layer
+length(layer_last_states)
+# 3
```
Comment (dfalbel) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:393]: It took me some time to understand the comment and the length here. Maybe print the `str` instead?

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd [hunk 15]
```diff
@@ -0,0 +1,757 @@
...
+
+# default args
+# return shape = (batch_size, units)
+#
+# note: there is an additional argument num_layers that we could use to specify a stacked LSTM - effectively composing two LSTM modules
+# default for num_layers is 1 though 
+lstm <- nn_lstm(
+  input_size = 1, # number of input features
+  hidden_size = 1, # number of hidden (and output!) features
+  batch_first = TRUE # for easy comparability with Keras
```
Comment (dfalbel) [inline _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:707]: This is not really really equivalent to keras, because keras has `bias=FALSE` by default. I know this doesn't change the point, so maybe just ignore that :)

**Comment by dfalbel** [comment]
Looks really great @skeydan ! Just left a few comments !

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd (hunk 13 above)
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:452]: makes sense!

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd (hunk 15 above)
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd:707]: oh, but Keras has `use_bias = TRUE` it seems?

File: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd (hunk 14 above)
Comment (skeydan) [reply _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd]: great idea, done!

**Comment by skeydan** [comment]
thanks a lot Daniel, great suggestions!
